integrations/access/opsgenie/app.go

```diff
@@ -17,15 +17,575 @@ limitations under the License.
 package opsgenie
 
 import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"google.golang.org/grpc"
+	grpcbackoff "google.golang.org/grpc/backoff"
+
+	tp "github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/client"
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
 	"github.com/gravitational/teleport/integrations/access/common"
+	"github.com/gravitational/teleport/integrations/access/common/teleport"
+	"github.com/gravitational/teleport/integrations/lib"
+	"github.com/gravitational/teleport/integrations/lib/backoff"
+	"github.com/gravitational/teleport/integrations/lib/credentials"
+	"github.com/gravitational/teleport/integrations/lib/logger"
+	"github.com/gravitational/teleport/integrations/lib/watcherjob"
 )
 
 const (
-	// opsgeniePluginName is used to tag Opsgenie GenericPluginData and as a Delegator in Audit log.
-	opsgeniePluginName = "opsgenie"
+	// pluginName is used to tag Opsgenie GenericPluginData and as a Delegator in Audit log.
+	pluginName = "opsgenie"
+	// minServerVersion is the minimal teleport version the plugin supports.
+	minServerVersion = "6.1.0"
+	// grpcBackoffMaxDelay is a maximum time GRPC client waits before reconnection attempt.
+	grpcBackoffMaxDelay = time.Second * 2
+	// initTimeout is used to bound execution time of health check and teleport version check.
+	initTimeout = time.Second * 10
+	// handlerTimeout is used to bound the execution time of watcher event handler.
+	handlerTimeout = time.Second * 5
+	// modifyPluginDataBackoffBase is an initial (minimum) backoff value.
+	modifyPluginDataBackoffBase = time.Millisecond
+	// modifyPluginDataBackoffMax is a backoff threshold
+	modifyPluginDataBackoffMax = time.Second
 )
 
+// Special kind of error that can be ignored.
+var errSkip = errors.New("")
+
+// App is a wrapper around the base app to allow for extra functionality.
+type App struct {
+	common.BaseApp
+
+	teleport teleport.Client
+	opsgenie *Client
+	mainJob  lib.ServiceJob
+	conf     Config
+}
+
 // NewOpsgenieApp initializes a new teleport-opsgenie app and returns it.
-func NewOpsgenieApp(conf *Config) *common.BaseApp {
-	return common.NewApp(conf, opsgeniePluginName)
+func NewOpsgenieApp(conf *Config) (*App, error) {
+	app := common.NewApp(conf, pluginName)
+	teleportClient, err := conf.GetTeleportClient(context.Background())
```

Comment:
We should accept `context.Context` here, or we should use `context.TODO()` instead of `context.Background()`.